File: mParticle-Apple-SDK/Location/MPLocationManager.m
#ifndef MPARTICLE_LOCATION_DISABLE
#import "MPLocationManager.h"
#import <UIKit/UIKit.h>

static BOOL _trackingLocation = NO;

#if TARGET_OS_IOS == 1
@interface MPLocationManager () <CLLocationManagerDelegate>
@end
#endif

@implementation MPLocationManager

#if TARGET_OS_IOS == 1
- (instancetype)initWithAccuracy:(CLLocationAccuracy)accuracy distanceFilter:(CLLocationDistance)distance authorizationRequest:(MPLocationAuthorizationRequest)authorizationRequest {
    self = [super init];
    
    CLAuthorizationStatus authorizationStatus = [CLLocationManager authorizationStatus];
    
    if (!self || authorizationStatus == kCLAuthorizationStatusRestricted || authorizationStatus == kCLAuthorizationStatusDenied) {
        return nil;
    }
    
    self.locationManager.desiredAccuracy = accuracy;
    self.locationManager.distanceFilter = distance;
    
    if ([[[UIDevice currentDevice] systemVersion] floatValue] >= 8.0) {
        NSDictionary *mainBundleInfoDictionary = [[NSBundle mainBundle] infoDictionary];
        
        if (authorizationRequest == MPLocationAuthorizationRequestAlways &&
            [self.locationManager respondsToSelector:@selector(requestAlwaysAuthorization)] &&
            mainBundleInfoDictionary[@"NSLocationAlwaysUsageDescription"])
        {
            [self.locationManager requestAlwaysAuthorization];
        } else if (authorizationRequest == MPLocationAuthorizationRequestWhenInUse &&
                   [self.locationManager respondsToSelector:@selector(requestWhenInUseAuthorization)] &&
                   mainBundleInfoDictionary[@"NSLocationWhenInUseUsageDescription"])
        {
            [self.locationManager requestWhenInUseAuthorization];
        } else {
            [self.locationManager startUpdatingLocation];
        }
    } else {
        [self.locationManager startUpdatingLocation];
    }
    
    _authorizationRequest = authorizationRequest;
    _requestedAccuracy = accuracy;
    _requestedDistanceFilter = distance;
    _trackingLocation = NO;
    _backgroundLocationTracking = YES;
    
    return self;
}

#pragma mark CLLocationManager Delegate
- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
    _trackingLocation = (status == kCLAuthorizationStatusAuthorizedAlways) || (status == kCLAuthorizationStatusAuthorizedWhenInUse);
    
    if (_trackingLocation) {
        [self.locationManager startUpdatingLocation];
    }
}

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-implementations"
- (void)locationManager:(CLLocationManager *)manager didUpdateToLocation:(CLLocation *)newLocation fromLocation:(CLLocation *)oldLocation {
    self.location = newLocation;
}
#pragma clang diagnostic pop

#pragma mark Public accessors
- (CLLocationManager *)locationManager {
    if (_locationManager) {
        return _locationManager;
    }
    
    CLAuthorizationStatus authorizationStatus = [CLLocationManager authorizationStatus];
    if (authorizationStatus == kCLAuthorizationStatusRestricted || authorizationStatus == kCLAuthorizationStatusDenied) {
        if (_locationManager) {
            _locationManager = nil;
            _location = nil;
            _trackingLocation = NO;
        }
        
        return nil;
    }
    
    [self willChangeValueForKey:@"locationManager"];
    _locationManager = [[CLLocationManager alloc] init];
    _locationManager.delegate = self;
    [self didChangeValueForKey:@"locationManager"];
    return _locationManager;
}

#pragma mark Public methods
- (void)endLocationTracking {
    [_locationManager stopUpdatingLocation];
    _locationManager = nil;
    
    _location = nil;
    _trackingLocation = NO;
}
#endif // #if TARGET_OS_IOS == 1

#pragma mark Class methods
+ (BOOL)trackingLocation {
    return _trackingLocation;
}

@end
#endif

File: mParticle-Apple-SDK/Location/MPLocationManager.swift
//
//  MPLocationManager.swift
//  mParticle-Apple-SDK
//
//  Created by Ben Baron on 9/12/24.
//

import Foundation
import UIKit

#if os(iOS) && !MPARTICLE_LOCATION_DISABLE
import CoreLocation
#endif

@objc final public class MPLocationManager_PRIVATE: NSObject {

    private static var _trackingLocation = false
    @objc public class var trackingLocation: Bool {
        return _trackingLocation
    }
    
    #if os(iOS) && !MPARTICLE_LOCATION_DISABLE
    private static var _locationManager: CLLocationManager?
    
    @objc public var location: CLLocation?
    @objc public private(set) var authorizationRequest: MPLocationAuthorizationRequest
    @objc public private(set) var requestedAccuracy: CLLocationAccuracy
    @objc public private(set) var requestedDistanceFilter: CLLocationDistance
    @objc public var backgroundLocationTracking: Bool

    @objc public var locationManager: CLLocationManager? {
        get {
            guard Self._locationManager == nil else {
                return Self._locationManager
            }
            
            let authorizationStatus = CLLocationManager.authorizationStatus()
            guard authorizationStatus != .restricted && authorizationStatus != .denied else {
                if let _ = Self._locationManager {
                    Self._locationManager = nil
                    location = nil
                    Self._trackingLocation = false
                }
                return nil
            }
            
            let _locationManager = CLLocationManager()
            _locationManager.delegate = self
            Self._locationManager = _locationManager
            return Self._locationManager
        }
    }
    
    @objc public init?(accuracy: CLLocationAccuracy, distanceFilter: CLLocationDistance, authorizationRequest: MPLocationAuthorizationRequest) {
        let authorizationStatus = CLLocationManager.authorizationStatus()
        guard authorizationStatus != .restricted && authorizationStatus != .denied else {
            return nil
        }
        
        self.authorizationRequest = authorizationRequest
        requestedAccuracy = accuracy
        requestedDistanceFilter = distanceFilter
        backgroundLocationTracking = true
        Self._trackingLocation = false
        super.init()
        
        // Must be run on the main thread or no delegate methods will be called
        DispatchQueue.main.async {
            if let locationManager = self.locationManager {
                locationManager.desiredAccuracy = accuracy
                locationManager.distanceFilter = distanceFilter
                
                let keys = Bundle.main.infoDictionary?.keys
                if let keys = keys, authorizationRequest == .always && keys.contains("NSLocationAlwaysUsageDescription") {
                    locationManager.requestAlwaysAuthorization()
                } else if let keys = keys, authorizationRequest == .whenInUse && keys.contains("NSLocationWhenInUseUsageDescription") {
                    locationManager.requestWhenInUseAuthorization()
                } else {
                    locationManager.startUpdatingLocation()
                }
            }
        }
    }

    @objc public func endLocationTracking() {
        Self._locationManager?.stopUpdatingLocation()
        Self._locationManager = nil
       
        location = nil
        Self._trackingLocation = false
    }
    
    #endif
}

#if os(iOS) && !MPARTICLE_LOCATION_DISABLE
extension MPLocationManager_PRIVATE: CLLocationManagerDelegate {
    public func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        Self._trackingLocation = (status == .authorizedAlways || status == .authorizedWhenInUse)
        
        if Self._trackingLocation {
            locationManager?.startUpdatingLocation()
        }
    }
        
    public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        location = locations.last
    }
}
#endif
